[PTD] Dump rcclexp proxy trace in pytorch

dmwu opened the pull request:
Summary:
Dump the active proxyOp status per rank and per communicator when WatchDog timeout or aborts.

Added 	
`#if defined(USE_ROCM) && defined(NCCL_COMM_DUMP)` guard in the print function, so only rcclexp users will see this dump in console.

This is the changes of the PTD.

Test Plan:
Job with A2A hang due to receiver failing to post receive operations https://fburl.com/mlhub/95vg12r3
 {F1971449692}

Reviewed By: c-p-i-o

Differential Revision: D67036093




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 commented:
Is ncclCommDump an API in RCCL? I am also not sure if torch should touch internals of NCCL such as proxy op.

dmwu commented:
> Is ncclCommDump an API in RCCL? I am also not sure if torch should touch internals of NCCL such as proxy op.

@kwen2501 Yes,  ncclCommDump is an API of our [internal fork of RCCL](https://www.internalfb.com/code/fbsource/[e78821f1a521]/third-party/rcclexp/develop/src/rccl.h?lines=836).  Torch calls this API to dump the RCCL proxy states (in strings) when rccl timeout or crashes. Torch itself should be agnostic of these states.

dmwu committed:
[PTD] Dump rcclexp proxy trace in pytorch (#143678)

Summary:

Dump the active proxyOp status per rank and per communicator when WatchDog timeout or aborts.

Added 	
`#if defined(USE_ROCM) && defined(NCCL_COMM_DUMP)` guard in the print function, so only rcclexp users will see this dump in console.

This is the changes of the PTD.

Test Plan:
Job with A2A hang due to receiver failing to post receive operations https://fburl.com/mlhub/95vg12r3
 {F1971449692}

Reviewed By: c-p-i-o

Differential Revision: D67036093
